src/irmin-pack/IO.ml

```diff
@@ -210,28 +240,60 @@ module Unix : S = struct
     | false ->
         let raw = raw ~mode ~version:current_version ~generation:0L file in
         v ~offset:0L ~version:current_version ~generation:0L raw
-    | true ->
+    | true -> (
         let x = Unix.openfile file Unix.[ O_EXCL; mode; O_CLOEXEC ] 0o644 in
         let raw = Raw.v x in
         if fresh then (
           let header =
             {
-              Raw.Header.version = current_version;
+              Raw.Header.version = bin_of_version current_version;
               offset = 0L;
               generation = 0L;
             }
           in
           Raw.Header.set raw header;
           v ~offset:0L ~version:current_version ~generation:0L raw)
         else
-          let { Raw.Header.version; offset; generation } = Raw.Header.get raw in
-          assert (version = current_version);
-          v ~offset ~version ~generation raw
+          let version = Raw.Version.get raw in
+          match version_of_bin version with
+          | Some `V1 ->
+              Log.debug (fun l -> l "[%s] file exists in V1" file);
+              if readonly then
+                invalid_arg
+                  "File is in version 1. Open the store in read-write mode to \
+                   migrate it to version 2";
+              let offset = Raw.Offset.get raw in
+              v ~offset ~version:`V1 ~generation:0L raw
+          | Some version ->
+              let { Raw.Header.offset; generation; _ } = Raw.Header.get raw in
+              v ~offset ~version ~generation raw
+          | None ->
+              let pp_full_version ppf v =
+                Fmt.pf ppf "%a (%S)" pp_version v (bin_of_version v)
+              in
+              Fmt.failwith "invalid version: got %S, expecting %a" version
+                Fmt.(Dump.list pp_full_version)
+                (List.map fst versions))
 
   let close t = Raw.close t.raw
-end
 
-let ( // ) = Filename.concat
+  let rmdir =
+    let rec aux path =
+      match Sys.is_directory path with
+      | true ->
+          Sys.readdir path
+          |> Array.iter (fun name -> aux (Filename.concat path name))
+      | false -> Sys.remove path
+    in
+    aux
+
+  let rename_dir ~src ~dst =
+    rmdir dst;
+    let cmd = Printf.sprintf "mv %s/* %s" src dst in
+    Fmt.epr "exec: %s\n%!" cmd;
+    let _ = Sys.command cmd in
+    Unix.rmdir src
```

Comment:
I believe the standard `rename` syscall works with directories.

Reply:
Indeed :)